### PR TITLE
Remove libspatialindex and add rtree PyPI package

### PR DESCRIPTION
**Pull request by JDBetteridge**
Requires https://github.com/firedrakeproject/libsupermesh/pull/4 to be merged and https://github.com/firedrakeproject/libspatialindex to be archived.

### CODE REVIEW

File: firedrake/mesh.py [hunk 1]
```diff
@@ -2254,38 +2256,50 @@ def _c_locator(self, tolerance=None):
...
-    """ % dict(geometric_dimension=self.geometric_dimension())
-
-            locator = compilation.load(src, "c", "locator",
-                                       cppargs=["-I%s" % os.path.dirname(__file__),
-                                                "-I%s/include" % sys.prefix]
-                                       + ["-I%s/include" % d for d in get_petsc_dir()],
-                                       ldargs=["-L%s/lib" % sys.prefix,
-                                               "-lspatialindex_c",
-                                               "-Wl,-rpath,%s/lib" % sys.prefix])
+            src += dedent(f"""
```
[the remark this answers is not in the document]
Comment (JDBetteridge) [reply firedrake/mesh.py:2259]: I 100% messed up the indentation the first time because of the triple quoted string had different indentation. We should use `dedent` throughout!

File: scripts/firedrake-install [hunk 2]
```diff
@@ -1254,7 +1254,7 @@ def build_and_install_h5py():
 
 def build_and_install_libsupermesh():
     log.info("Installing libsupermesh")
-    url = "git+https://github.com/firedrakeproject/libsupermesh.git"
+    url = "git+https://github.com/firedrakeproject/libsupermesh.git@JDBetteridge/remove_libspatialindex"
```
Comment (JDBetteridge) [inline scripts/firedrake-install:1257]:
```suggestion
    url = "git+https://github.com/firedrakeproject/libsupermesh.git"
```

File: scripts/firedrake-install [hunk 3]
```diff
@@ -1266,18 +1266,17 @@ def build_and_install_libsupermesh():
...
+                    "cmake", "..", "-DBUILD_SHARED_LIBS=ON",
+                    "-DCMAKE_INSTALL_PREFIX=" + firedrake_env,
+                    "-DMPI_C_COMPILER=" + cc,
+                    "-DMPI_CXX_COMPILER=" + cxx,
+                    "-DMPI_Fortran_COMPILER=" + f90,
+                    "-DCMAKE_Fortran_COMPILER=" + f90,
+                    "-DMPIEXEC_EXECUTABLE=" + args.mpiexec,
+                    "-DLIBSUPERMESH_AUTO_COMPILER_FLAGS=OFF"
+                ]
+                print(" ".join(cmd))
```
Comment (connorjward) [inline scripts/firedrake-install:1279]:
```suggestion
```

File: scripts/firedrake-install [hunk 4]
```diff
@@ -1266,18 +1266,17 @@ def build_and_install_libsupermesh():
...
             with directory("build"):
-                cmd = ["cmake", "..", "-DBUILD_SHARED_LIBS=ON",
-                       "-DCMAKE_INSTALL_PREFIX=" + firedrake_env,
-                       "-DMPI_C_COMPILER=" + cc,
-                       "-DMPI_CXX_COMPILER=" + cxx,
-                       "-DMPI_Fortran_COMPILER=" + f90,
-                       "-DCMAKE_Fortran_COMPILER=" + f90,
-                       "-DLIBSUPERMESH_AUTO_COMPILER_FLAGS=OFF"]
-                fortran, version = sniff_compiler(f90)
-                if fortran == "GNU" and version >= Version("10"):
```
Comment (connorjward) [inline scripts/firedrake-install:1277]: I assume this is safe to remove?

File: scripts/firedrake-install [hunk 5]
```diff
@@ -1934,16 +1900,19 @@ if mode == "install":
...
-
-    build_and_install_libspatialindex()
 
     if "loopy" in packages:
         # We do want to install loopy's dependencies.
         install("loopy/")
         packages.remove("loopy")
 
     with pipargs("--no-deps"):
+        try:
```
Comment (connorjward) [inline scripts/firedrake-install:1910]: Why is this here?

File: scripts/firedrake-install [hunk 6]
```diff
@@ -1952,13 +1921,20 @@ if mode == "install":
                 install(p+"/")
             sys.path.append(os.getcwd() + "/" + p)
 
+    with environment(**compiler_env):
+        build_and_install_libsupermesh()
+
+    with pipargs("--no-deps"):
```
Comment (connorjward) [inline scripts/firedrake-install:1927]: Put the context managers together? https://stackoverflow.com/questions/893333/multiple-variables-in-a-with-statement

File: scripts/firedrake-install [hunk 7]
```diff
@@ -2113,6 +2087,14 @@ Please consider updating your PETSc manually.
             else:
                 install(p+"/")
 
+        with environment(**compiler_env):
+            build_and_install_libsupermesh()
+
+        with pipargs("--no-deps"):
```
Comment (connorjward) [inline scripts/firedrake-install:2093]: As above.

**Review by connorjward — CHANGES_REQUESTED** [review]
Only minor comments. Otherwise looks good.

File: scripts/firedrake-install (hunk 4 above)
Comment (JDBetteridge) [reply scripts/firedrake-install:1277]: This is already in the libsupermesh `CMakeLists.txt`

**Review by dham — APPROVED** [review]

File: scripts/firedrake-install (hunk 5 above)
Comment (JDBetteridge) [reply scripts/firedrake-install:1910]: Needs to be removed for reordering, needs a try because sometimes the packages aren't in the list.

**Review by connorjward — APPROVED** [review]

**Review by ksagiyam — APPROVED** [review]
